### PR TITLE
Improve catalog function interface

### PR DESCRIPTION
**Pull request by andyfengHKU**
# Description

This is a followed-up PR of #4726. We now only expose FunctionEntry from catalog and hide FunctionCatalogSet. The same thing should be done for Macro but I kinda think it should first be moved as a separate CatalogSet. Right now in FunctionCatalogSet we have both `FunctionCatalogEntry` & `MacroCatalogEntry`.  Either they should be the same CatalogEntry or they belong to different CatalogSet. 

@acquamarin & @ray6080 we should discuss this.

Also @acquamarin can u test what will happen if we start doing
```
CALL macro_function() RETURN *
CALL scalar_function() RETURN *
RETURN table/gds_function()
```
I suspect we don't throw properly in some of these edge cases.

Fixes # (issue)

# Contributor agreement

- [ ] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

**Review by ray6080 — APPROVED** [review]

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -358,29 +363,19 @@ void Catalog::addFunction(Transaction* transaction, CatalogEntryType entryType,
 }
 
 void Catalog::dropFunction(Transaction* transaction, const std::string& name) {
-    const auto entry = functions->getEntry(transaction, name);
-    if (entry == nullptr) {
+    if (!containsFunction(transaction, name)) {
```
Comment (acquamarin) [inline src/catalog/catalog.cpp:366]: should binder always call containsFunction first before calling dropFunction?

**Review by acquamarin — APPROVED** [review]

**Comment by acquamarin** [comment]
I tested the following:
1. CALL macro function
```
CREATE MACRO add5(x) AS x + 5;
CALL ADD5(3) RETURN *;
```
2. CALL scalar function
```
CALL SIN(5) RETURN * => Error: `Binder exception: sin is not a table or algorithm function.`
```

3. RETURN table function
```
return show_tables(); => Error: Binder exception: SHOW_TABLES is a TABLE_FUNCTION_ENTRY. Scalar function, aggregate function or macro was expected. 
```
